iscsi/iscsi.go

```diff
@@ -78,15 +78,17 @@ type Connector struct {
 	DoCHAPDiscovery bool `json:"do_chap_discovery"`
 }
 
+var version string = "1.0.0"
+
 func init() {
-	// by default we don't log anything, EnableDebugLogging() can turn on some tracing
-	debug = log.New(ioutil.Discard, "", 0)
+	// By default we produce info and error log entries which can be redirected using EnableDebugLogging()
+	klog.InfoS("[] csi-lib-iscsi", "version", version)
 }
 
-// EnableDebugLogging provides a mechanism to turn on debug logging for this package
-// output is written to the provided io.Writer
+// EnableDebugLogging provides a mechanism to allow a caller to set their own io.Writer
 func EnableDebugLogging(writer io.Writer) {
-	debug = log.New(writer, "DEBUG: ", log.Ldate|log.Ltime|log.Lshortfile)
+	klog.InfoS("Enable a caller's io.Writer")
```

Comment:
This entire function should be replaced with a mechanism that allows a user of the library to pass in a logr.Logger instance or a context.

Then the library should log through that instance or use `klog.FromContext` to retrieve a logger instance that it uses.

Reply:
Okay, should I undertake that change, or remove this function. If we use FromContext, they may allow main() application verbosity levels to passed into the library and then allow additional levels of logging filtered by verbosity level - such as klog.V(1).InfoS("Verbose"), klog.V(3).InfoS("Debug"),klog.V(4).InfoS("Trace").

Reply:
Remove the function. A library should not modify any global state, not even when called to do so.

For contextual logging, never use klog directly. Always log through a logger instance.

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#contextual-logging-in-kubernetes

Reply:
I modified the library to pass in ctx context.Context, based on my understanding of the :preferred approach". With these changes, the logging verbosity level is passed into the library and allows the user to control the verbosity from the command line using -v=<level>. Tested these changes out using our CSI driver and a pvc-hell test written by Enix.